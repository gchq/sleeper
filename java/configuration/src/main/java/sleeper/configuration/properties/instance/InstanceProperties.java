/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.configuration.properties.instance;

import com.amazonaws.services.s3.AmazonS3;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.configuration.properties.SleeperProperties;
import sleeper.configuration.properties.SleeperPropertyIndex;
import sleeper.configuration.properties.format.SleeperPropertiesPrettyPrinter;

import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Properties;

import static sleeper.configuration.properties.PropertiesUtils.loadProperties;
import static sleeper.configuration.properties.instance.CdkDefinedInstanceProperty.CONFIG_BUCKET;
import static sleeper.configuration.properties.instance.CommonProperty.TAGS;

/**
 * Contains all the properties needed to deploy an instance of Sleeper.
 */
public class InstanceProperties extends SleeperProperties<InstanceProperty> {
    private static final Logger LOGGER = LoggerFactory.getLogger(InstanceProperties.class);

    public static final String S3_INSTANCE_PROPERTIES_FILE = "instance.properties";

    protected Map<String, String> tags = new HashMap<>();

    public InstanceProperties() {
        super();
    }

    public InstanceProperties(Properties properties) {
        super(properties);
        tags = csvTagsToMap(get(TAGS));
    }

    public static InstanceProperties copyOf(InstanceProperties instanceProperties) {
        return new InstanceProperties(loadProperties(instanceProperties.saveAsString()));
    }

    @Override
    protected void init() {
        tags = csvTagsToMap(get(TAGS));
        super.init();
    }

    public Map<String, String> getTags() {
        return tags;
    }

    public void setTags(Map<String, String> tagsMap) {
        tags.clear();
        tags.putAll(tagsMap);
        set(TAGS, tagsToString(tags));
    }

    public void loadTags(Properties tagsProperties) {
        tags.clear();
        tagsProperties.stringPropertyNames().forEach(tagName -> tags.put(tagName, tagsProperties.getProperty(tagName)));
        if (!tags.isEmpty()) {
            set(TAGS, tagsToString(tags));
        }
    }

    @Override
    public String get(InstanceProperty property) {
        return compute(property, value -> property.computeValue(value, this));
    }

    public Properties getTagsProperties() {
        Properties tagsProperties = new Properties();
        tags.forEach(tagsProperties::setProperty);
        return tagsProperties;
    }

    public String getTagsPropertiesAsString() throws IOException {
        StringWriter stringWriter = new StringWriter();
        Properties tagsProperties = getTagsProperties();
        tagsProperties.store(stringWriter, "");
        return stringWriter.toString();
    }

    public static String getConfigBucketFromInstanceId(String instanceId) {
        return String.join("-", "sleeper", instanceId, "config").toLowerCase(Locale.ROOT);
    }

    public void loadFromS3GivenInstanceId(AmazonS3 s3Client, String instanceId) {
        String configBucket = getConfigBucketFromInstanceId(instanceId);
        loadFromS3(s3Client, configBucket);
    }

    public void loadFromS3(AmazonS3 s3Client, String bucket) {
        super.loadFromS3(s3Client, bucket, S3_INSTANCE_PROPERTIES_FILE);
    }

    public static Properties loadPropertiesFromS3GivenInstanceId(AmazonS3 s3Client, String instanceId) {
        return loadProperties(loadStringFromS3GivenInstanceId(s3Client, instanceId));
    }

    private static String loadStringFromS3GivenInstanceId(AmazonS3 s3Client, String instanceId) {
        String configBucket = getConfigBucketFromInstanceId(instanceId);
        return s3Client.getObjectAsString(configBucket, S3_INSTANCE_PROPERTIES_FILE);
    }

    public void saveToS3(AmazonS3 s3Client) {
        super.saveToS3(s3Client, get(CONFIG_BUCKET), S3_INSTANCE_PROPERTIES_FILE);
        LOGGER.info("Saved instance properties to bucket {}, key {}", get(CONFIG_BUCKET), S3_INSTANCE_PROPERTIES_FILE);
    }

    @Override
    public SleeperPropertyIndex<InstanceProperty> getPropertiesIndex() {
        return InstanceProperty.Index.INSTANCE;
    }

    @Override
    protected SleeperPropertiesPrettyPrinter<InstanceProperty> getPrettyPrinter(PrintWriter writer) {
        return SleeperPropertiesPrettyPrinter.forInstanceProperties(writer);
    }

    public static Map<String, String> csvTagsToMap(String csvTags) {
        Map<String, String> tags = new HashMap<>();
        if (null != csvTags && !csvTags.isEmpty()) {
            String[] split = csvTags.split(",");
            for (int i = 0; i < split.length; i += 2) {
                tags.put(split[i], split[i + 1]);
            }
        }
        return tags;
    }

    public static String tagsToString(Map<String, String> tags) {
        StringBuilder builder = new StringBuilder();
        int count = 0;
        for (Map.Entry<String, String> entry : tags.entrySet()) {
            if (count > 0) {
                builder.append(",");
            }
            builder.append(entry.getKey());
            builder.append(",");
            builder.append(entry.getValue());
            count++;
        }
        return builder.toString();
    }
}
