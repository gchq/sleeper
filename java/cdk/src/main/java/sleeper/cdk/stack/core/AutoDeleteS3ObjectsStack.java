/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.cdk.stack.core;

import software.amazon.awscdk.CustomResource;
import software.amazon.awscdk.Duration;
import software.amazon.awscdk.NestedStack;
import software.amazon.awscdk.customresources.Provider;
import software.amazon.awscdk.services.lambda.IFunction;
import software.amazon.awscdk.services.logs.ILogGroup;
import software.amazon.awscdk.services.logs.LogGroup;
import software.amazon.awscdk.services.s3.Bucket;
import software.amazon.awscdk.services.s3.IBucket;
import software.constructs.Construct;

import sleeper.cdk.jars.BuiltJars;
import sleeper.cdk.jars.LambdaCode;
import sleeper.cdk.stack.core.LoggingStack.LogGroupRef;
import sleeper.cdk.util.Utils;
import sleeper.core.deploy.LambdaHandler;
import sleeper.core.properties.instance.InstanceProperties;
import sleeper.core.util.EnvironmentUtils;

import java.util.Map;

public class AutoDeleteS3ObjectsStack extends NestedStack {

    private IFunction lambda;
    private Provider provider;

    public AutoDeleteS3ObjectsStack(Construct scope, String id, InstanceProperties instanceProperties, BuiltJars jars, LoggingStack loggingStack) {
        super(scope, id);
        createLambda(id, instanceProperties, jars, loggingStack.getLogGroup(LogGroupRef.AUTO_DELETE_S3_OBJECTS),
                loggingStack.getLogGroup(LogGroupRef.AUTO_DELETE_S3_OBJECTS_PROVIDER));
    }

    public AutoDeleteS3ObjectsStack(Construct scope, String id, InstanceProperties instanceProperties, BuiltJars jars, Integer retentionDays) {
        super(scope, id);
        ILogGroup logGroup = LogGroup.Builder.create(this, id + "-AutoDeleteLambdaLogGroup")
                .logGroupName("s3-bucket-autodelete")
                .retention(Utils.getRetentionDays(retentionDays))
                .build();
        ILogGroup providerLogGroup = LogGroup.Builder.create(this, id + "-AutoDeleteProviderLogGroup")
                .logGroupName("s3-bucket-autodelete-provider")
                .retention(Utils.getRetentionDays(retentionDays))
                .build();
        createLambda(id, instanceProperties, jars, logGroup, providerLogGroup);
    }

    private void createLambda(String id, InstanceProperties instanceProperties, BuiltJars jars, ILogGroup logGroup, ILogGroup providerLogGroup) {

        // Jars bucket
        IBucket jarsBucket = Bucket.fromBucketName(this, "JarsBucket", jars.bucketName());
        LambdaCode lambdaCode = jars.lambdaCode(jarsBucket);

        String functionName = String.join("-", "sleeper",
                Utils.cleanInstanceId(instanceProperties), "auto-delete-s3-objects");

        lambda = lambdaCode.buildFunction(this, LambdaHandler.AUTO_DELETE_S3_OBJECTS, id + "Lambda", builder -> builder
                .functionName(functionName)
                .memorySize(2048)
                .environment(EnvironmentUtils.createDefaultEnvironmentNoConfigBucket(instanceProperties))
                .description("Lambda for auto-deleting S3 objects")
                .logGroup(logGroup)
                .timeout(Duration.minutes(10)));

        provider = Provider.Builder.create(this, id + "Provider")
                .onEventHandler(lambda)
                .logGroup(providerLogGroup)
                .build();
    }

    public void grantAccessToCustomResource(String id, InstanceProperties instanceProperties,
            IBucket bucket, String bucketName) {

        bucket.grantRead(lambda);
        bucket.grantDelete(lambda);

        CustomResource.Builder.create(this, id + "-AutoDelete")
                .resourceType("Custom::AutoDeleteS3Objects")
                .properties(Map.of("bucket", bucketName))
                .serviceToken(provider.getServiceToken())
                .build();
    }

}
