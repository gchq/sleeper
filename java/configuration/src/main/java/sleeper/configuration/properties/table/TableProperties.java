/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.configuration.properties.table;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.configuration.properties.SleeperProperties;
import sleeper.configuration.properties.SleeperPropertiesValidationReporter;
import sleeper.configuration.properties.SleeperPropertyIndex;
import sleeper.configuration.properties.format.SleeperPropertiesPrettyPrinter;
import sleeper.configuration.properties.instance.InstanceProperties;
import sleeper.core.schema.Schema;
import sleeper.core.schema.SchemaSerDe;
import sleeper.core.table.TableStatus;

import java.io.PrintWriter;
import java.util.Optional;
import java.util.Properties;

import static sleeper.configuration.properties.PropertiesUtils.loadProperties;
import static sleeper.configuration.properties.table.TableProperty.COMPACTION_FILES_BATCH_SIZE;
import static sleeper.configuration.properties.table.TableProperty.SCHEMA;
import static sleeper.configuration.properties.table.TableProperty.STATESTORE_CLASSNAME;
import static sleeper.configuration.properties.table.TableProperty.TABLE_ID;
import static sleeper.configuration.properties.table.TableProperty.TABLE_NAME;
import static sleeper.configuration.properties.table.TableProperty.TABLE_ONLINE;

public class TableProperties extends SleeperProperties<TableProperty> {
    private static final Logger LOGGER = LoggerFactory.getLogger(TableProperties.class);

    // Schema is cached for faster access
    private Schema schema;
    private final InstanceProperties instanceProperties; // Used for default properties

    public TableProperties(InstanceProperties instanceProperties) {
        this.instanceProperties = instanceProperties;
    }

    public TableProperties(InstanceProperties instanceProperties, Properties properties) {
        super(properties);
        this.instanceProperties = instanceProperties;
        schema = loadSchema(properties);
    }

    public static TableProperties copyOf(TableProperties tableProperties) {
        InstanceProperties instanceProperties = InstanceProperties.copyOf(tableProperties.instanceProperties);
        return new TableProperties(instanceProperties, loadProperties(tableProperties.saveAsString()));
    }

    public static TableProperties loadAndValidate(InstanceProperties instanceProperties, Properties properties) {
        TableProperties tableProperties = new TableProperties(instanceProperties, properties);
        tableProperties.validate();
        return tableProperties;
    }

    public static TableProperties reinitialise(TableProperties tableProperties, Properties newProperties) {
        return new TableProperties(tableProperties.instanceProperties, newProperties);
    }

    private static Schema loadSchema(Properties properties) {
        return Optional.ofNullable(properties.getProperty(SCHEMA.getPropertyName()))
                .map(Schema::loadFromString)
                .orElse(null);
    }

    @Override
    protected void init() {
        String schemaProperty = get(TableProperty.SCHEMA);
        if (schemaProperty != null) {
            schema = Schema.loadFromString(schemaProperty);
        }
        super.init();
    }

    @Override
    public void validate(SleeperPropertiesValidationReporter reporter) {
        super.validate(reporter);

        // This limit is based on calls to WriteTransactItems in DynamoDBFileReferenceStore.atomicallyUpdateX.
        // Also see the DynamoDB documentation:
        // https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/transaction-apis.html
        if ("sleeper.statestore.dynamodb.DynamoDBStateStore".equals(get(STATESTORE_CLASSNAME))
                && getInt(COMPACTION_FILES_BATCH_SIZE) > 49) {
            LOGGER.warn("Detected a compaction batch size for this table which would be incompatible with the " +
                    "chosen statestore. Maximum value is 49.");
            reporter.invalidProperty(COMPACTION_FILES_BATCH_SIZE, get(COMPACTION_FILES_BATCH_SIZE));
        }
    }

    @Override
    public String get(TableProperty property) {
        return compute(property, value -> property.computeValue(value, instanceProperties, this));
    }

    public Schema getSchema() {
        return schema;
    }

    public void setSchema(Schema schema) {
        this.schema = schema;
        set(TableProperty.SCHEMA, new SchemaSerDe().toJson(schema));
    }

    @Override
    public SleeperPropertyIndex<TableProperty> getPropertiesIndex() {
        return TableProperty.Index.INSTANCE;
    }

    @Override
    protected SleeperPropertiesPrettyPrinter<TableProperty> getPrettyPrinter(PrintWriter writer) {
        return SleeperPropertiesPrettyPrinter.forTableProperties(writer);
    }

    public TableStatus getStatus() {
        return TableStatus.uniqueIdAndName(get(TABLE_ID), get(TABLE_NAME), getBoolean(TABLE_ONLINE));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        TableProperties that = (TableProperties) o;

        return new EqualsBuilder()
                .appendSuper(super.equals(o))
                .append(instanceProperties, that.instanceProperties)
                .isEquals();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder(17, 37)
                .appendSuper(super.hashCode())
                .append(instanceProperties)
                .toHashCode();
    }
}
