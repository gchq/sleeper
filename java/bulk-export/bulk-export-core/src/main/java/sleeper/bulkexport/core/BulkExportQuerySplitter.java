/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.bulkexport.core;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.bulkexport.core.model.BulkExportLeafPartitionQuery;
import sleeper.bulkexport.core.model.BulkExportQuery;
import sleeper.core.partition.Partition;
import sleeper.core.partition.PartitionTree;
import sleeper.core.properties.table.TableProperties;
import sleeper.core.statestore.StateStore;
import sleeper.core.statestore.StateStoreException;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.function.Supplier;
import java.util.stream.Collectors;

import static sleeper.core.properties.table.TableProperty.QUERY_PROCESSOR_CACHE_TIMEOUT;

/**
 * Splits up an export query into leaf partition export queries.
 */
public class BulkExportQuerySplitter {
    private static final Logger LOGGER = LoggerFactory.getLogger(BulkExportQuerySplitter.class);

    private final StateStore stateStore;
    private final TableProperties tableProperties;
    private final Supplier<String> idSupplier;
    private final Supplier<Instant> timeSupplier;
    private List<Partition> leafPartitions;
    private PartitionTree partitionTree;
    private Map<String, List<String>> partitionToFiles;
    private Instant nextInitialiseTime;

    public BulkExportQuerySplitter(TableProperties tableProperties, StateStore stateStore) {
        this(stateStore, tableProperties, () -> UUID.randomUUID().toString(), Instant::now);
    }

    public BulkExportQuerySplitter(StateStore stateStore, TableProperties tableProperties, Supplier<String> idSupplier, Supplier<Instant> timeSupplier) {
        this.stateStore = stateStore;
        this.tableProperties = tableProperties;
        this.idSupplier = idSupplier;
        this.timeSupplier = timeSupplier;
        init(timeSupplier.get());
    }

    /**
     * Initialises the partitions and the mapping from partitions to active files if
     * needed.
     *
     * @throws StateStoreException if the state store can't be accessed
     */
    public void initIfNeeded() throws StateStoreException {
        Instant now = timeSupplier.get();
        if (nextInitialiseTime.isAfter(now)) {
            LOGGER.debug("Not refreshing state for table {}", tableProperties.getStatus());
            return;
        }
        init(now);
    }

    private void init(Instant now) throws StateStoreException {
        List<Partition> partitions = stateStore.getAllPartitions();
        Map<String, List<String>> partitionToFileMapping = stateStore.getPartitionToReferencedFilesMap();
        init(partitions, partitionToFileMapping, now);
    }

    private void init(List<Partition> partitions, Map<String, List<String>> partitionToFileMapping, Instant now) {
        leafPartitions = partitions.stream()
                .filter(Partition::isLeafPartition)
                .collect(Collectors.toList());
        partitionTree = new PartitionTree(partitions);
        partitionToFiles = partitionToFileMapping;
        nextInitialiseTime = now.plus(tableProperties.getInt(QUERY_PROCESSOR_CACHE_TIMEOUT), ChronoUnit.SECONDS);
        LOGGER.info("Loaded state for table {}. Found {} partitions. Next initialise time: {}",
                tableProperties.getStatus(), partitions.size(), nextInitialiseTime);
    }

    /**
     * Splits up a query into a sub-query per relevant leaf partition. For each leaf partition, it
     * finds the parent partitions in the tree and adds any files still belonging
     * to the parent to the sub query.
     *
     * @param  bulkExportQuery the query to be split up
     * @return                 a list of {@link LeafPartitionQuery}s
     */
    public List<BulkExportLeafPartitionQuery> splitIntoLeafPartitionQueries(BulkExportQuery bulkExportQuery) {
        LOGGER.debug("There are {} relevant leaf partitions", leafPartitions.size());

        List<BulkExportLeafPartitionQuery> leafPartitionQueriesList = new ArrayList<>();
        for (Partition partition : leafPartitions) {
            List<String> files = getFiles(partition);

            if (files.isEmpty()) {
                LOGGER.info("No files for partition {}", partition.getId());
                continue;
            }

            // For each leaf partition, create query with pre-populated list of files that
            // need to be read.
            BulkExportLeafPartitionQuery bulkExportLeafPartitionQuery = BulkExportLeafPartitionQuery
                    .forPartition(bulkExportQuery, tableProperties, partition)
                    .subExportId(idSupplier.get())
                    .files(files)
                    .build();
            LOGGER.debug("Created {}", bulkExportLeafPartitionQuery);
            leafPartitionQueriesList.add(bulkExportLeafPartitionQuery);
        }

        return leafPartitionQueriesList;
    }

    /**
     * Gets a list of file paths for a partition.
     *
     * @param  partition which partition to get the files from
     * @return           a list of files
     */
    protected List<String> getFiles(Partition partition) {
        // Get all partitions up to the root of the tree
        List<String> relevantPartitions = new ArrayList<>();
        relevantPartitions.add(partition.getId());
        relevantPartitions.addAll(partitionTree.getAllAncestorIds(partition.getId()));

        // Get relevant files
        List<String> files = new ArrayList<>();
        for (String partitionId : relevantPartitions) {
            List<String> filesForPartition = partitionToFiles.get(partitionId);
            if (null != filesForPartition) {
                files.addAll(partitionToFiles.get(partitionId));
            }
        }
        return files;
    }
}
