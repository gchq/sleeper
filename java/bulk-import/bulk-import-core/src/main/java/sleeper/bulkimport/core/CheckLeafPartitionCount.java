/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.bulkimport.core;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.bulkimport.core.job.BulkImportJob;
import sleeper.core.properties.table.TableProperties;
import sleeper.core.statestore.StateStore;

import static sleeper.core.properties.table.TableProperty.BULK_IMPORT_MIN_LEAF_PARTITION_COUNT;

public class CheckLeafPartitionCount {
    private static final Logger LOGGER = LoggerFactory.getLogger(CheckLeafPartitionCount.class);

    private CheckLeafPartitionCount() {
    }

    public static boolean hasMinimumPartitions(
            TableProperties tableProperties, StateStore stateStore, BulkImportJob job) {
        int leafPartitionCount = stateStore.getLeafPartitions().size();
        int minPartitionCount = tableProperties.getInt(BULK_IMPORT_MIN_LEAF_PARTITION_COUNT);
        if (leafPartitionCount < minPartitionCount) {
            LOGGER.info("Minimum partition count was {}, but found {} leaf partitions. Skipping job {}",
                    minPartitionCount, leafPartitionCount, job.getId());
            return false;
        } else {
            LOGGER.info("Minimum partition count has been met. Running job {}", job.getId());
            return true;
        }
    }
}
