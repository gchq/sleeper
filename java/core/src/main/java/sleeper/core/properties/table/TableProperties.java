/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.properties.table;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;

import sleeper.core.properties.PropertyGroup;
import sleeper.core.properties.SleeperProperties;
import sleeper.core.properties.SleeperPropertiesPrettyPrinter;
import sleeper.core.properties.SleeperPropertiesValidationReporter;
import sleeper.core.properties.SleeperPropertyIndex;
import sleeper.core.properties.instance.InstanceProperties;
import sleeper.core.schema.Schema;
import sleeper.core.schema.SchemaSerDe;
import sleeper.core.table.TableStatus;

import java.io.PrintWriter;
import java.util.Optional;
import java.util.Properties;
import java.util.stream.Collectors;

import static sleeper.core.properties.PropertiesUtils.loadProperties;
import static sleeper.core.properties.table.TableProperty.SCHEMA;
import static sleeper.core.properties.table.TableProperty.TABLE_ID;
import static sleeper.core.properties.table.TableProperty.TABLE_NAME;
import static sleeper.core.properties.table.TableProperty.TABLE_ONLINE;

/**
 * Contains values of the properties to configure a Sleeper table.
 */
public class TableProperties extends SleeperProperties<TableProperty> {

    // Schema is cached for faster access
    private Schema schema;
    private final InstanceProperties instanceProperties; // Used for default properties

    public TableProperties(InstanceProperties instanceProperties) {
        this.instanceProperties = instanceProperties;
    }

    public TableProperties(InstanceProperties instanceProperties, Properties properties) {
        super(properties);
        this.instanceProperties = instanceProperties;
        schema = loadSchema(properties);
    }

    /**
     * Creates a copy of the given table properties.
     *
     * @param  tableProperties the table properties
     * @return                 the copy
     */
    public static TableProperties copyOf(TableProperties tableProperties) {
        InstanceProperties instanceProperties = InstanceProperties.copyOf(tableProperties.instanceProperties);
        return new TableProperties(instanceProperties, loadProperties(tableProperties.saveAsString()));
    }

    /**
     * Creates and validates an instance of this class with the given property values.
     *
     * @param  instanceProperties the instance properties
     * @param  properties         the property values
     * @return                    the table properties
     */
    public static TableProperties createAndValidate(InstanceProperties instanceProperties, Properties properties) {
        TableProperties tableProperties = new TableProperties(instanceProperties, properties);
        tableProperties.validate();
        return tableProperties;
    }

    /**
     * Creates an instance with different property values than the given instance. Performs no validation, and uses the
     * same instance properties as the given instance.
     *
     * @param  tableProperties the original table properties
     * @param  newProperties   the property values
     * @return                 the new table properties
     */
    public static TableProperties recreateWithoutValidation(TableProperties tableProperties, Properties newProperties) {
        return new TableProperties(tableProperties.instanceProperties, newProperties);
    }

    private static Schema loadSchema(Properties properties) {
        return Optional.ofNullable(properties.getProperty(SCHEMA.getPropertyName()))
                .map(Schema::loadFromString)
                .orElse(null);
    }

    @Override
    protected void init() {
        String schemaProperty = get(TableProperty.SCHEMA);
        if (schemaProperty != null) {
            schema = Schema.loadFromString(schemaProperty);
        }
        super.init();
    }

    @Override
    public void validate(SleeperPropertiesValidationReporter reporter) {
        super.validate(reporter);
    }

    @Override
    public String get(TableProperty property) {
        return compute(property, value -> property.computeValue(value, instanceProperties, this));
    }

    public Schema getSchema() {
        return schema;
    }

    /**
     * Sets the schema of the Sleeper table.
     *
     * @param schema the schema
     */
    public void setSchema(Schema schema) {
        this.schema = schema;
        set(TableProperty.SCHEMA, new SchemaSerDe().toJson(schema));
    }

    @Override
    public SleeperPropertyIndex<TableProperty> getPropertiesIndex() {
        return TableProperty.Index.INSTANCE;
    }

    @Override
    protected SleeperPropertiesPrettyPrinter<TableProperty> getPrettyPrinter(PrintWriter writer) {
        return createPrettyPrinter(writer);
    }

    /**
     * Creates a printer to be used to display all table properties.
     *
     * @param  writer the writer to write to
     * @return        the pretty printer
     */
    public static SleeperPropertiesPrettyPrinter<TableProperty> createPrettyPrinter(PrintWriter writer) {
        return SleeperPropertiesPrettyPrinter.builder()
                .properties(TableProperty.getAll(), TablePropertyGroup.getAll())
                .writer(writer).build();
    }

    /**
     * Creates a printer to be used to display table properties in a given group.
     *
     * @param  writer the writer to write to
     * @param  group  the group to display
     * @return        the pretty printer
     */
    public static SleeperPropertiesPrettyPrinter<TableProperty> createPrettyPrinterWithGroup(
            PrintWriter writer, PropertyGroup group) {
        return SleeperPropertiesPrettyPrinter.builder()
                .sortedProperties(TableProperty.getAll().stream()
                        .filter(property -> property.getPropertyGroup().equals(group))
                        .collect(Collectors.toList()))
                .writer(writer).hideUnknownProperties(true).build();
    }

    public TableStatus getStatus() {
        return TableStatus.uniqueIdAndName(get(TABLE_ID), get(TABLE_NAME), getBoolean(TABLE_ONLINE));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        TableProperties that = (TableProperties) o;

        return new EqualsBuilder()
                .appendSuper(super.equals(o))
                .append(instanceProperties, that.instanceProperties)
                .isEquals();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder(17, 37)
                .appendSuper(super.hashCode())
                .append(instanceProperties)
                .toHashCode();
    }
}
