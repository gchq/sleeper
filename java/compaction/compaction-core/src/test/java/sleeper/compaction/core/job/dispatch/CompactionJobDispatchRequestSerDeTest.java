/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.compaction.core.job.dispatch;

import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.junit.jupiter.api.Test;

import sleeper.core.properties.instance.InstanceProperties;
import sleeper.core.properties.table.TableProperties;

import java.time.Instant;

import static org.assertj.core.api.Assertions.assertThat;
import static sleeper.core.properties.instance.CdkDefinedInstanceProperty.DATA_BUCKET;
import static sleeper.core.properties.table.TableProperty.TABLE_ID;
import static sleeper.core.properties.testutils.InstancePropertiesTestHelper.createTestInstanceProperties;
import static sleeper.core.properties.testutils.TablePropertiesTestHelper.createTestTablePropertiesWithNoSchema;

public class CompactionJobDispatchRequestSerDeTest {

    CompactionJobDispatchRequestSerDe serDe = new CompactionJobDispatchRequestSerDe();
    InstanceProperties instanceProperties = createTestInstanceProperties();
    TableProperties tableProperties = createTestTablePropertiesWithNoSchema(instanceProperties);

    @Test
    void shouldConvertRequestToFromJson() {
        // Given
        instanceProperties.set(DATA_BUCKET, "test-bucket");
        tableProperties.set(TABLE_ID, "test-table");
        CompactionJobDispatchRequest request = CompactionJobDispatchRequest.forTableWithBatchIdAtTime(
                tableProperties, "test-batch", Instant.parse("2024-11-18T12:01:00Z"));

        // When
        String json = serDe.toJsonPrettyPrint(request);

        // Then
        assertThat(serDe.fromJson(json)).isEqualTo(request);
        Approvals.verify(json, new Options().forFile().withExtension(".json"));
    }
}
