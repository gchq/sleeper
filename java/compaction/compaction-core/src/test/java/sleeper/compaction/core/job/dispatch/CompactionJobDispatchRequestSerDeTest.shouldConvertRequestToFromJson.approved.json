{
  "tableId": "test-table",
  "batchKey": "s3a://test-bucket/test-table/compactions/test-batch.json",
  "expiryTime": 1731931350000
}